client.go

```diff
@@ -0,0 +1,287 @@
+package rendezvous
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p-rendezvous/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+	host "github.com/libp2p/go-libp2p-host"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+)
+
+var (
+	DiscoverAsyncInterval = 2 * time.Minute
+)
+
+type Rendezvous interface {
+	Register(ctx context.Context, ns string, ttl int) error
+	Unregister(ctx context.Context, ns string) error
+	Discover(ctx context.Context, ns string, limit int, cookie []byte) ([]Registration, []byte, error)
+	DiscoverAsync(ctx context.Context, ns string) (<-chan Registration, error)
+}
+
+type Registration struct {
+	Peer pstore.PeerInfo
+	Ns   string
+	Ttl  int
+}
+
+func NewRendezvousClient(host host.Host, rp peer.ID) Rendezvous {
+	return &client{
+		host: host,
+		rp:   rp,
+	}
+}
+
+type client struct {
+	host host.Host
+	rp   peer.ID
+}
+
+func (cli *client) Register(ctx context.Context, ns string, ttl int) error {
+	s, err := cli.host.NewStream(ctx, cli.rp, RendezvousProto)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	req := newRegisterMessage(ns, pstore.PeerInfo{ID: cli.host.ID(), Addrs: cli.host.Addrs()}, ttl)
+	err = w.WriteMsg(req)
+	if err != nil {
+		return err
+	}
+
+	var res pb.Message
+	err = r.ReadMsg(&res)
+	if err != nil {
+		return err
+	}
+
+	if res.GetType() != pb.Message_REGISTER_RESPONSE {
+		return fmt.Errorf("Unexpected response: %s", res.GetType().String())
+	}
+
+	status := res.GetRegisterResponse().GetStatus()
+	if status != pb.Message_OK {
+		return RendezvousError{Status: status, Text: res.GetRegisterResponse().GetStatusText()}
+	}
+
+	return nil
+}
+
+func Register(ctx context.Context, rz Rendezvous, ns string, ttl int) error {
+	if ttl < 120 {
+		return fmt.Errorf("registration TTL is too short")
+	}
+
+	err := rz.Register(ctx, ns, ttl)
+	if err != nil {
+		return err
+	}
+
+	go registerRefresh(ctx, rz, ns, ttl)
+	return nil
+}
+
+func registerRefresh(ctx context.Context, rz Rendezvous, ns string, ttl int) {
```

Comment:
It doesnt appear that this gets closed down when unregister is called. So if i call register then unregister, this routine would re-register me.

Reply:
hrm, yes, the two are not correlated; you'd have to cancel the context to stop registration refresh.

Reply:
Note that this is a utility function to augment the interface with persistent registrations.